platform/android/src/native_map_view.cpp

```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
-    if (mode == MapObserver::CameraChangeMode::Immediate) {
-        notifyMapChange(MapChange::MapChangeRegionWillChange);
-    } else {
-        notifyMapChange(MapChange::MapChangeRegionWillChangeAnimated);
-    }
+    static auto onCameraWillChange = javaClass.GetMethod<void(jboolean)>(*_env, "onCameraWillChange");
+    javaPeer->Call(*_env, onCameraWillChange, (jboolean) (mode == MapObserver::CameraChangeMode::Animated));
 }
 
+/**
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onCameraIsChanging().
+ *
+ * May be called from any thread
+ */
 void NativeMapView::onCameraIsChanging() {
-    notifyMapChange(MapChange::MapChangeRegionIsChanging);
+    assert(vm != nullptr);
+    android::UniqueEnv _env = android::AttachEnv();
+    static auto onCameraIsChanging = javaClass.GetMethod<void()>(*_env, "onCameraIsChanging");
+    javaPeer->Call(*_env, onCameraIsChanging);
 }
 
+/**
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onCameraDidChange().
+ *
+ * May be called from any thread
+ */
 void NativeMapView::onCameraDidChange(MapObserver::CameraChangeMode mode) {
-    if (mode == MapObserver::CameraChangeMode::Immediate) {
-        notifyMapChange(MapChange::MapChangeRegionDidChange);
-    } else {
-        notifyMapChange(MapChange::MapChangeRegionDidChangeAnimated);
-    }
+    assert(vm != nullptr);
+    android::UniqueEnv _env = android::AttachEnv();
+    static auto onCameraDidChange = javaClass.GetMethod<void(jboolean)>(*_env, "onCameraDidChange");
+    javaPeer->Call(*_env, onCameraDidChange, (jboolean) (mode == MapObserver::CameraChangeMode::Animated));
 }
 
+/**
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onWillStartLoadingMap().
+ *
+ * May be called from any thread
+ */
 void NativeMapView::onWillStartLoadingMap() {
-    notifyMapChange(MapChange::MapChangeWillStartLoadingMap);
+    assert(vm != nullptr);
+    android::UniqueEnv _env = android::AttachEnv();
+    static auto onWillStartLoadingMap = javaClass.GetMethod<void()>(*_env, "onWillStartLoadingMap");
+    javaPeer->Call(*_env, onWillStartLoadingMap);
 }
 
+/**
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onDidFinishLoadingMap().
+ *
+ * May be called from any thread
+ */
 void NativeMapView::onDidFinishLoadingMap() {
-    notifyMapChange(MapChange::MapChangeDidFinishLoadingMap);
+    assert(vm != nullptr);
+    android::UniqueEnv _env = android::AttachEnv();
+    static auto onDidFinishLoadingMap = javaClass.GetMethod<void()>(*_env, "onDidFinishLoadingMap");
+    javaPeer->Call(*_env, onDidFinishLoadingMap);
 }
 
-void NativeMapView::onDidFailLoadingMap(std::exception_ptr) {
-    notifyMapChange(MapChange::MapChangeDidFailLoadingMap);
+/**
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onDidFailLoadingMap().
+ *
+ * May be called from any thread
+ */
+void NativeMapView::onDidFailLoadingMap(std::exception_ptr eptr) {
+    assert(vm != nullptr);
+    try {
+        std::rethrow_exception(eptr);
```

Comment:
Do we want to throw an exception here? Seems like we always want to invoke the listener. If re-throwing is required, it should happen where the exception is produced.